File: skills/.archived/brain-skills-go/fix_fences.go
package main

import (
	"flag"
	"fmt"
	"io/fs"
	"os"
	"path/filepath"
	"regexp"
	"strings"
)

// FixFencesOutput represents the output for fix-fences command
type FixFencesOutput struct {
	Success      bool     `json:"success"`
	FixedFiles   []string `json:"fixedFiles"`
	TotalFixed   int      `json:"totalFixed"`
	TotalScanned int      `json:"totalScanned"`
	DryRun       bool     `json:"dryRun"`
	Error        string   `json:"error,omitempty"`
}

var (
	openingPattern = regexp.MustCompile("^(\\s*)```(\\w+)")
	closingPattern = regexp.MustCompile("^(\\s*)```\\s*$")
)

func runFixFences() error {
	// Use FlagSet for subcommand to avoid conflicts with global flags
	fs := flag.NewFlagSet("fix-fences", flag.ContinueOnError)
	pattern := fs.String("pattern", "**/*.md", "Glob pattern for markdown files")
	dryRun := fs.Bool("dry-run", false, "Report changes without writing files")
	help := fs.Bool("help", false, "Show usage information")

	// Parse args after the subcommand name
	if err := fs.Parse(os.Args[2:]); err != nil {
		return err
	}

	if *help {
		fmt.Fprintln(os.Stderr, "fix-fences - Fix malformed markdown code fence closings")
		fmt.Fprintln(os.Stderr, "")
		fmt.Fprintln(os.Stderr, "Usage:")
		fmt.Fprintln(os.Stderr, "  brain-skills fix-fences [flags] [directories...]")
		fmt.Fprintln(os.Stderr, "")
		fmt.Fprintln(os.Stderr, "Flags:")
		fmt.Fprintln(os.Stderr, "  --pattern string   Glob pattern for markdown files (default \"**/*.md\")")
		fmt.Fprintln(os.Stderr, "  --dry-run          Report changes without writing files")
		fmt.Fprintln(os.Stderr, "  --help             Show this help message")
		return nil
	}

	directories := fs.Args()
	if len(directories) == 0 {
		directories = []string{"."}
	}

	output := FixFencesOutput{
		Success:    true,
		FixedFiles: []string{},
		DryRun:     *dryRun,
	}

	for _, dir := range directories {
		info, err := os.Stat(dir)
		if err != nil {
			if os.IsNotExist(err) {
				fmt.Fprintf(os.Stderr, "Warning: directory does not exist: %s\n", dir)
				continue
			}
			output.Success = false
			output.Error = fmt.Sprintf("failed to access directory %s: %v", dir, err)
			return outputJSON(output)
		}
		if !info.IsDir() {
			fmt.Fprintf(os.Stderr, "Warning: not a directory: %s\n", dir)
			continue
		}

		absDir, err := filepath.Abs(dir)
		if err != nil {
			output.Success = false
			output.Error = fmt.Sprintf("failed to resolve path %s: %v", dir, err)
			return outputJSON(output)
		}

		fixed, scanned, err := fixMarkdownFiles(absDir, *pattern, *dryRun)
		if err != nil {
			output.Success = false
			output.Error = fmt.Sprintf("failed to process directory %s: %v", dir, err)
			return outputJSON(output)
		}

		output.FixedFiles = append(output.FixedFiles, fixed...)
		output.TotalScanned += scanned
	}

	output.TotalFixed = len(output.FixedFiles)
	return outputJSON(output)
}

func fixMarkdownFences(content string) string {
	lines := strings.Split(content, "\n")
	result := make([]string, 0, len(lines))
	inCodeBlock := false
	blockIndent := ""

	for _, line := range lines {
		openingMatch := openingPattern.FindStringSubmatch(line)
		closingMatch := closingPattern.FindStringSubmatch(line)

		if openingMatch != nil {
			if inCodeBlock {
				result = append(result, blockIndent+"```")
			}
			result = append(result, line)
			blockIndent = openingMatch[1]
			inCodeBlock = true
		} else if closingMatch != nil {
			result = append(result, line)
			inCodeBlock = false
			blockIndent = ""
		} else {
			result = append(result, line)
		}
	}

	if inCodeBlock {
		result = append(result, blockIndent+"```")
	}

	return strings.Join(result, "\n")
}

func fixMarkdownFiles(directory, pattern string, dryRun bool) ([]string, int, error) {
	var fixed []string
	scanned := 0

	err := filepath.WalkDir(directory, func(path string, d fs.DirEntry, err error) error {
		if err != nil {
			return err
		}

		if d.IsDir() {
			return nil
		}

		matched, err := matchPattern(path, directory, pattern)
		if err != nil {
			return err
		}
		if !matched {
			return nil
		}

		scanned++

		content, err := os.ReadFile(path)
		if err != nil {
			return fmt.Errorf("failed to read %s: %w", path, err)
		}

		fixedContent := fixMarkdownFences(string(content))

		if string(content) != fixedContent {
			if !dryRun {
				if err := os.WriteFile(path, []byte(fixedContent), 0644); err != nil {
					return fmt.Errorf("failed to write %s: %w", path, err)
				}
			}
			fixed = append(fixed, path)
		}

		return nil
	})

	if err != nil {
		return nil, 0, err
	}

	return fixed, scanned, nil
}

func matchPattern(path, baseDir, pattern string) (bool, error) {
	relPath, err := filepath.Rel(baseDir, path)
	if err != nil {
		return false, err
	}

	if strings.Contains(pattern, "**") {
		parts := strings.Split(pattern, "**")
		if len(parts) == 2 {
			suffix := strings.TrimPrefix(parts[1], "/")
			suffix = strings.TrimPrefix(suffix, string(filepath.Separator))
			if suffix == "" {
				return true, nil
			}
			matched, err := filepath.Match(suffix, filepath.Base(path))
			if err != nil {
				return false, err
			}
			return matched, nil
		}
	}

	matched, err := filepath.Match(pattern, relPath)
	if err != nil {
		return false, err
	}
	if matched {
		return true, nil
	}

	matched, err = filepath.Match(pattern, filepath.Base(path))
	if err != nil {
		return false, err
	}
	return matched, nil
}

File: templates/skills/fix-markdown-fences/fix_fences.ts
#!/usr/bin/env bun
/**
 * Fix malformed markdown code fence closings.
 *
 * Scans markdown files and repairs closing fences that incorrectly include
 * language identifiers (```python instead of ```).
 *
 * EXIT CODES (ADR-035):
 *    0 - Success: Fences fixed or no issues found
 */

import { Glob } from "bun";

const OPENING_PATTERN = /^(\s*)```(\w+)/;
const CLOSING_PATTERN = /^(\s*)```\s*$/;

function repairMarkdownFences(content: string): string {
  const lines = content.split(/\r?\n/);
  const result: string[] = [];
  let inCodeBlock = false;
  let codeBlockIndent = "";

  for (const line of lines) {
    const openingMatch = line.match(OPENING_PATTERN);
    const closingMatch = line.match(CLOSING_PATTERN);

    if (openingMatch) {
      if (inCodeBlock) {
        result.push(codeBlockIndent + "```");
      }
      result.push(line);
      codeBlockIndent = openingMatch[1];
      inCodeBlock = true;
    } else if (closingMatch) {
      result.push(line);
      inCodeBlock = false;
      codeBlockIndent = "";
    } else {
      result.push(line);
    }
  }

  if (inCodeBlock) {
    result.push(codeBlockIndent + "```");
  }

  return result.join("\n");
}

async function fixFences(
  directories: string[],
  pattern: string = "*.md",
): Promise<number> {
  let totalFixed = 0;

  for (const directory of directories) {
    const dirFile = Bun.file(directory);
    const stat = await dirFile.exists();
    if (!stat) {
      console.error(`Warning: Directory does not exist: ${directory}`);
      continue;
    }

    const glob = new Glob(`**/${pattern}`);

    for await (const path of glob.scan({ cwd: directory })) {
      const fullPath = `${directory}/${path}`;
      const file = Bun.file(fullPath);

      let content: string;
      try {
        content = await file.text();
      } catch {
        continue;
      }

      if (!content) {
        continue;
      }

      const fixedContent = repairMarkdownFences(content);
      if (content !== fixedContent) {
        await Bun.write(fullPath, fixedContent);
        totalFixed += 1;
      }
    }
  }

  return totalFixed;
}

function parseArgs(): { directories: string[]; pattern: string } {
  const args = process.argv.slice(2);
  let directories: string[] = ["."];
  let pattern = "*.md";

  let i = 0;
  while (i < args.length) {
    if (args[i] === "--directories") {
      directories = [];
      i++;
      while (i < args.length && !args[i].startsWith("--")) {
        directories.push(args[i]);
        i++;
      }
    } else if (args[i] === "--pattern") {
      i++;
      if (i < args.length) {
        pattern = args[i];
        i++;
      }
    } else if (args[i] === "--help" || args[i] === "-h") {
      console.log("Usage: fix_fences.ts [--directories dir1 dir2 ...] [--pattern *.md]");
      console.log("\nFix malformed markdown code fence closings");
      console.log("\nOptions:");
      console.log("  --directories  Directories to scan (default: [\".\"])");
      console.log("  --pattern      File pattern to match (default: *.md)");
      process.exit(0);
    } else {
      i++;
    }
  }

  return { directories, pattern };
}

async function main(): Promise<number> {
  const { directories, pattern } = parseArgs();
  const totalFixed = await fixFences(directories, pattern);

  if (totalFixed === 0) {
    console.log("No files needed fixing");
  } else {
    console.log(`\nTotal: fixed ${totalFixed} file(s)`);
  }

  return 0;
}

process.exit(await main());
